File: backend/src/main/java/com/lsoft/aguapotable/service/TarifaService.java
package com.lsoft.aguapotable.service;

import com.lsoft.aguapotable.domain.entity.Comunidad;
import com.lsoft.aguapotable.domain.entity.Tarifa;
import com.lsoft.aguapotable.domain.entity.TarifaRango;
import com.lsoft.aguapotable.domain.repository.ComunidadRepository;
import com.lsoft.aguapotable.domain.repository.TarifaRepository;
import com.lsoft.aguapotable.dto.request.TarifaRequest;
import com.lsoft.aguapotable.exception.BusinessException;
import com.lsoft.aguapotable.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class TarifaService {

    private final TarifaRepository tarifaRepository;
    private final ComunidadRepository comunidadRepository;

    public List<Tarifa> listarActivas(Long comunidadId) {
        return tarifaRepository.findByComunidadIdAndActivoTrue(comunidadId);
    }

    public Tarifa obtenerPorId(Long id, Long comunidadId) {
        Tarifa tarifa = tarifaRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Tarifa", id));
        if (!tarifa.getComunidad().getId().equals(comunidadId)) {
            throw new ResourceNotFoundException("Tarifa", id);
        }
        return tarifa;
    }

    @Transactional
    public Tarifa crear(TarifaRequest request, Long comunidadId) {
        Comunidad comunidad = comunidadRepository.findById(comunidadId)
                .orElseThrow(() -> new ResourceNotFoundException("Comunidad", comunidadId));

        Tarifa tarifa = Tarifa.builder()
                .nombre(request.getNombre())
                .comunidad(comunidad)
                .tipoSuscriptor(request.getTipoSuscriptor())
                .cuotaFija(request.getCuotaFija())
                .porcentajeMora(request.getPorcentajeMora())
                .diasGracia(request.getDiasGracia())
                .vigenciaDesde(request.getVigenciaDesde())
                .vigenciaHasta(request.getVigenciaHasta())
                .activo(true)
                .rangos(new ArrayList<>())
                .build();

        if (request.getRangos() != null) {
            for (TarifaRequest.RangoRequest rango : request.getRangos()) {
                TarifaRango tr = TarifaRango.builder()
                        .tarifa(tarifa)
                        .rangoDesde(rango.getRangoDesde())
                        .rangoHasta(rango.getRangoHasta())
                        .precioPorM3(rango.getPrecioPorM3())
                        .build();
                tarifa.getRangos().add(tr);
            }
        }

        return tarifaRepository.save(tarifa);
    }

    @Transactional
    public Tarifa actualizar(Long id, TarifaRequest request, Long comunidadId) {
        Tarifa tarifa = obtenerPorId(id, comunidadId);

        tarifa.setNombre(request.getNombre());
        tarifa.setTipoSuscriptor(request.getTipoSuscriptor());
        tarifa.setCuotaFija(request.getCuotaFija());
        tarifa.setPorcentajeMora(request.getPorcentajeMora());
        tarifa.setDiasGracia(request.getDiasGracia());
        tarifa.setVigenciaDesde(request.getVigenciaDesde());
        tarifa.setVigenciaHasta(request.getVigenciaHasta());

        // Reemplazar rangos (orphanRemoval se encarga de borrar los viejos)
        tarifa.getRangos().clear();
        if (request.getRangos() != null) {
            for (TarifaRequest.RangoRequest rango : request.getRangos()) {
                TarifaRango tr = TarifaRango.builder()
                        .tarifa(tarifa)
                        .rangoDesde(rango.getRangoDesde())
                        .rangoHasta(rango.getRangoHasta())
                        .precioPorM3(rango.getPrecioPorM3())
                        .build();
                tarifa.getRangos().add(tr);
            }
        }

        return tarifaRepository.save(tarifa);
    }

    @Transactional
    public void desactivar(Long id, Long comunidadId) {
        Tarifa tarifa = obtenerPorId(id, comunidadId);
        tarifa.setActivo(false);
        tarifaRepository.save(tarifa);
    }
}

File: backend/src/tarifa/tarifa.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Tarifa, TarifaRango, Comunidad } from '../common/entities';
import { TarifaRequest } from './dto/tarifa.dto';
import { ResourceNotFoundException } from '../common/exceptions';

@Injectable()
export class TarifaService {
  constructor(
    @InjectRepository(Tarifa)
    private readonly repo: Repository<Tarifa>,
    @InjectRepository(Comunidad)
    private readonly comunidadRepo: Repository<Comunidad>,
    @InjectRepository(TarifaRango)
    private readonly rangoRepo: Repository<TarifaRango>,
  ) {}

  async listarActivas(comunidadId: number): Promise<Tarifa[]> {
    return this.repo.find({
      where: { comunidadId, activo: true },
      relations: ['rangos'],
    });
  }

  async obtenerPorId(id: number, comunidadId: number): Promise<Tarifa> {
    const tarifa = await this.repo.findOne({
      where: { id },
      relations: ['rangos'],
    });
    if (!tarifa) throw new ResourceNotFoundException('Tarifa', id);
    if (tarifa.comunidadId !== Number(comunidadId)) {
      throw new ResourceNotFoundException('Tarifa', id);
    }
    return tarifa;
  }

  async crear(dto: TarifaRequest, comunidadId: number): Promise<Tarifa> {
    const comunidad = await this.comunidadRepo.findOne({
      where: { id: comunidadId },
    });
    if (!comunidad)
      throw new ResourceNotFoundException('Comunidad', comunidadId);

    const tarifa = this.repo.create({
      nombre: dto.nombre,
      comunidadId,
      tipoSuscriptor: dto.tipoSuscriptor,
      cuotaFija: String(dto.cuotaFija),
      porcentajeMora: String(dto.porcentajeMora ?? 0.02),
      diasGracia: dto.diasGracia ?? 15,
      vigenciaDesde: dto.vigenciaDesde ?? null,
      vigenciaHasta: dto.vigenciaHasta ?? null,
      activo: true,
      rangos: [],
    });

    if (dto.rangos) {
      tarifa.rangos = dto.rangos.map((r) =>
        this.rangoRepo.create({
          rangoDesde: r.rangoDesde,
          rangoHasta: r.rangoHasta ?? null,
          precioPorM3: String(r.precioPorM3),
        }),
      );
    }

    return this.repo.save(tarifa);
  }

  async actualizar(
    id: number,
    dto: TarifaRequest,
    comunidadId: number,
  ): Promise<Tarifa> {
    const tarifa = await this.obtenerPorId(id, comunidadId);

    tarifa.nombre = dto.nombre;
    tarifa.tipoSuscriptor = dto.tipoSuscriptor;
    tarifa.cuotaFija = String(dto.cuotaFija);
    tarifa.porcentajeMora = String(dto.porcentajeMora ?? 0.02);
    tarifa.diasGracia = dto.diasGracia ?? 15;
    tarifa.vigenciaDesde = dto.vigenciaDesde ?? null;
    tarifa.vigenciaHasta = dto.vigenciaHasta ?? null;

    // Remove old rangos
    if (tarifa.rangos && tarifa.rangos.length > 0) {
      await this.rangoRepo.remove(tarifa.rangos);
    }

    tarifa.rangos = [];
    if (dto.rangos) {
      tarifa.rangos = dto.rangos.map((r) =>
        this.rangoRepo.create({
          rangoDesde: r.rangoDesde,
          rangoHasta: r.rangoHasta ?? null,
          precioPorM3: String(r.precioPorM3),
          tarifaId: tarifa.id,
        }),
      );
    }

    return this.repo.save(tarifa);
  }

  async desactivar(id: number, comunidadId: number): Promise<void> {
    const tarifa = await this.obtenerPorId(id, comunidadId);
    tarifa.activo = false;
    await this.repo.save(tarifa);
  }
}
